httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/MutualGssSchemeBase.java

```diff
@@ -0,0 +1,365 @@
+            }
+            if (gsse.getMajor() == GSSException.DEFECTIVE_TOKEN
+                    || gsse.getMajor() == GSSException.DUPLICATE_TOKEN
+                    || gsse.getMajor() == GSSException.OLD_TOKEN) {
+                throw new AuthenticationException(gsse.getMessage(), gsse);
+            }
+            // other error
+            throw new AuthenticationException(gsse.getMessage(), gsse);
+        }
+    }
+
+    protected GSSManager getManager() {
+        return GSSManager.getInstance();
+    }
+
+    /**
+     * @since 4.4
+     */
+    protected byte[] generateGSSToken(
+            final byte[] input, final Oid oid, final String gssServiceName, final String gssHostname) throws GSSException {
+        final GSSManager manager = getManager();
+        final GSSName peerName = manager.createName(gssServiceName + "@" + gssHostname, GSSName.NT_HOSTBASED_SERVICE);
+
+        if (gssContext == null) {
+            gssContext = createGSSContext(manager, oid, peerName, gssCredential);
+        }
+        if (input != null) {
+            return gssContext.initSecContext(input, 0, input.length);
+        }
+        return gssContext.initSecContext(new byte[] {}, 0, 0);
+    }
+
+    /**
+     * @since 5.0
+     */
+    protected GSSContext createGSSContext(
+            final GSSManager manager,
+            final Oid oid,
+            final GSSName peerName,
+            final GSSCredential gssCredential) throws GSSException {
+        final GSSContext gssContext = manager.createContext(peerName.canonicalize(oid), oid, gssCredential,
+                GSSContext.DEFAULT_LIFETIME);
+        gssContext.requestMutualAuth(mutualAuth);
+        gssContext.requestCredDeleg(config.isRequestDelegCreds());
+        return gssContext;
+    }
+
+    /**
+     * @since 4.4
+     */
+    protected abstract byte[] generateToken(byte[] input, String gssServiceName, String gssHostname) throws GSSException;
+
+    @Override
+    public boolean isChallengeComplete() {
+        // For the mutual authentication response, this is should technically return true.
+        // However, the HttpAuthenticator immediately fails the authentication
+        // process if we return true, so we only return true here if the authentication has failed.
+        return this.state == State.FAILED;
+    }
+
+    @Override
+    public boolean isChallengeExpected() {
+        return state == State.TOKEN_SENT && mutualAuth;
+    }
+
+    @Override
+    public boolean isResponseReady(
+            final HttpHost host,
+            final CredentialsProvider credentialsProvider,
+            final HttpContext context) throws AuthenticationException {
+
+        Args.notNull(host, "Auth host");
+        Args.notNull(credentialsProvider, "CredentialsProvider");
+
+        setGssCredential(credentialsProvider, host, context);
+        return true;
+    }
+
+    protected void setGssCredential(final CredentialsProvider credentialsProvider,
```

Comment:
This is the fix for the bug I found, @michael-o .
When we specifx an explicit kerberos crential (as opposed to relying on the jaas default)
Then it was only set from isResponseReady() which is too late, as we need it earlier, in processChallenge().
I had to add code to processChallange to process the credential, so that we can use it there.